2018 Robot/src/org/usfirst/frc/team4999/robot/sensors/VMXPi.java

```diff
@@ -13,61 +13,62 @@
  */
 public class VMXPi implements PIDSource {
 
-    private static int PORT = 5800;
+  private static int PORT = 5800;
+  private static int PACKET_LENGTH = 48;
+
 	private DatagramSocket server;
-	DatagramPacket packet = new DatagramPacket(new byte[17], 17);
-	
+	DatagramPacket packet = new DatagramPacket(new byte[PACKET_LENGTH], PACKET_LENGTH);
+
 	private Thread recieveLoop;
-	
-	private double zAngle, zRate;
-	
+
+	private double zAngle, zRate, pitch, roll, xVelocity, yVelocity;
+	private double zOffset;
+
 	// When two synchronized blocks are synchronized on the same object, only one block is allowed to run at a time.
 	// By synchronizing before we read and write the zAngle and zRates, we can make sure they don't change while we're reading them
 	private Object syncLock = new Object();
-	
+
 	private PIDSourceType m_source = PIDSourceType.kDisplacement;
-	
+
 	private long millis;
-    
+
 	private static VMXPi instance;
-	
+
 	public static VMXPi getInstance() {
 		if(instance == null) {
 			instance = new VMXPi();
 		}
 		return instance;
 	}
-	
+
 	private VMXPi() {
 		try {
 			server = new DatagramSocket(PORT);
 		} catch (IOException e) {
 			e.printStackTrace();
 		}
 		recieveLoop = new Thread() {
-			ByteBuffer buff = ByteBuffer.allocate(17);
-			
+
+		ByteBuffer buff = ByteBuffer.allocate(PACKET_LENGTH);
+
+
 			@Override
 			public void run() {
 				setName("VMX Thread");
 				while(!Thread.interrupted()) {
 					try {
 						server.receive(packet);
-						byte[] data = packet.getData();
-						int checksum = 0;
-						for(int i = 1; i < data.length; i++) {
-							checksum += (data[i]&0xFF);
+						buff.rewind();
+						buff.put(packet.getData());
```

Comment:
`buff` is only ever accessed from the thread constantly reading  data from the socket. The `syncLock` is so that the `zAngle`, `zRate`, `pitch`, `roll`, `xVelocity`, `yVelocity`, and `millis` aren't read while they're being overwritten. I'm not sure if java would let the variables be read while they're being written, but I decided to err on the side of caution